File: src/cmd/fo/fo.c
#include "../../stacks/C.h"
#include "../../main/C.h"
#include "../../misc/C.h"

#define bcopy(s, d, n) memcpy(d, s, n)
#define bzero(s, n) memset(s, 0, n)

/* Table of constant values */

static int c__1 = 1;

void fo_()
{
    /* System generated locals */
    int i__1, i__2;

    /* Local variables */
    int nfld;
    short nwts;
    int ichan;
    float refwt;
    float sigwt;
    float totwt;
    int iabfld;
    int nrefof, nsigof;
    short itmplc[NWEIGHTS];
    float tempwt[NWEIGHTS];

#include "fo.dc.h"

/* call parse to advance the ntty pointer. */
#include "fo.ec.h"

    if (coreFull[0] == 0) {
	error_("No data in stack 1");
    }
    if ((stk_[0].nprc & 4) != 0) {
	error_("Stack already folded");
    }
    if (stk_[0].refch == (float)-9999.) {
	error_("NO FOLD - not freq switched");
    }
    nfld = round(stk_[0].expch - stk_[0].refch);
    iabfld = ABS(nfld);
    if (stk_[0].numpt - iabfld < 5) {
	error_("NO FOLD - too few chans left");
    }
    if (nfld > 0) {
	nsigof = iabfld;
	nrefof = 0;
	stk_[0].expch -= nsigof;
	stk_[0].refch -= nsigof;
    } else {
	nsigof = 0;
	nrefof = iabfld;
    }
    bzero(tempwt, sizeof(tempwt));
    bzero(itmplc, sizeof(itmplc));
    nwts = 0;

    stk_[0].numpt -= iabfld;
    i__1 = stk_[0].numpt;
    for (ichan = 1; ichan <= i__1; ++ichan) {
	i__2 = ichan + nsigof;
	sigwt = wt_(&i__2, &c__1);
	i__2 = ichan + nrefof;
	refwt = wt_(&i__2, &c__1);
	totwt = sigwt + refwt;
	stk_[0].stak[ichan - 1] = (stk_[0].stak[ichan + nsigof - 1] * sigwt - 
		stk_[0].stak[ichan + nrefof - 1] * refwt) / totwt;
	pkwts_(&totwt, &ichan, tempwt, itmplc, &nwts);
/* L100: */
    }
    bcopy(tempwt, stk_[0].wght, sizeof(tempwt));
    bcopy(itmplc, stk_[0].locwt, sizeof(itmplc));
    stk_[0].nwts = nwts;
    bzero(&stk_[0].stak[stk_[0].numpt], iabfld << 2);
    stk_[0].rms *= (float).707106781;
    stk_[0].nprc |= 4;
    main_.pltd = 0;
} /* fo_ */

File: src/cmd/fo/fo.f
	subroutine fo
	implicit undefined (a-z)
	include '../../stacks/F.h'
	include '../../main/F.h'
	include 'fo.df.h'
	real*4 tempwt(15),totwt,sigwt,refwt
	integer nsigof,nrefof,iabfld,nfld,ichan,irond
	integer*2 itmplc(15),nwts
	automatic nsigof,nrefof,iabfld,nfld,itmplc,nwts,ichan
     1	,tempwt,totwt,sigwt,refwt
c functions
	integer stfull
	real*8 wt
	include 'fo.if.h'

c call parse to advance the ntty pointer.
	include 'fo.ef.h'

	if(stfull(1).eq.0) call error("No data in stack 1")
	if(and(nprc1,4).ne.0)call error("Stack already folded")
	if(refch1.eq.notfs) call error("NO FOLD - not freq switched")
	nfld=irond(expch1-refch1)
	iabfld=iabs(nfld)
	if(numpt1-iabfld.lt.5)call error("NO FOLD - too few chans left")
	if(nfld.gt.0) then
		nsigof=iabfld
		nrefof=0
		expch1=expch1-nsigof
		refch1=refch1-nsigof
	else
		nsigof=0
		nrefof=iabfld
	endif
	call clear(tempwt(1),60)
	call clear(itmplc(1),30)
	nwts=0
	numpt1=numpt1-iabfld
	do 100 ichan=1,numpt1
		sigwt=wt(ichan+nsigof,1)
		refwt=wt(ichan+nrefof,1)
		totwt=sigwt+refwt
		stak1(ichan)=(stak1(ichan+nsigof)*sigwt
     1			-stak1(ichan+nrefof)*refwt)/totwt
		call pkwts(totwt,ichan,tempwt,itmplc,nwts)
100	continue
	call move(tempwt,wght1,60)
	call move(itmplc,locwt1,30)
	nwts1=nwts
	call clear(stak1(numpt1+1),4*iabfld)
	rms1=rms1*0.707106781
	nprc1=or(nprc1,4)
	pltd=0
	end
